storage/src/key_value/agent/block.rs

```diff
@@ -0,0 +1,364 @@
+
+        self.inner.delete(KeyValueColumn::BlockHeader, &hash[..])?;
+
+        self.inner.delete(KeyValueColumn::BlockTransactions, &hash[..])?;
+
+        let mut child_hashes = self.get_child_block_hashes(&header.previous_block_hash)?;
+
+        if let Some(index) = child_hashes.iter().position(|x| x == hash) {
+            child_hashes.remove(index);
+
+            self.update_child_block_hashes(&header.previous_block_hash, &child_hashes[..])?;
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn get_block_hash(&mut self, block_num: u32) -> Result<Option<Digest>> {
+        let hash = self
+            .inner
+            .get(KeyValueColumn::BlockIndex, &block_num.to_le_bytes()[..])?
+            .map(|x| x[..].into());
+        Ok(hash)
+    }
+
+    fn get_block_hash_guarded(&mut self, block_num: u32) -> Result<Digest> {
+        let hash = self
+            .inner
+            .get(KeyValueColumn::BlockIndex, &block_num.to_le_bytes()[..])?
+            .map(|x| x[..].into())
+            .ok_or_else(|| anyhow!("missing canon hash"))?;
+        Ok(hash)
+    }
+
+    pub(super) fn get_block_header(&mut self, hash: &Digest) -> Result<SerialBlockHeader> {
+        let header = self
+            .inner
+            .get(KeyValueColumn::BlockHeader, &hash[..])?
+            .ok_or_else(|| anyhow!("block header missing"))?;
+        let header = SerialBlockHeader::read(&mut &header[..])?;
+        Ok(header)
+    }
+
+    pub(super) fn get_block_state(&mut self, hash: &Digest) -> Result<BlockStatus> {
+        let index = self.inner.get(KeyValueColumn::BlockIndex, &hash[..])?;
+        match index {
+            Some(index) => {
+                let block_number = Self::read_u32(&index[..])?;
+                Ok(BlockStatus::Committed(block_number as usize))
+            }
+            None => {
+                if self.inner.exists(KeyValueColumn::BlockHeader, &hash[..])? {
+                    Ok(BlockStatus::Uncommitted)
+                } else {
+                    Ok(BlockStatus::Unknown)
+                }
+            }
+        }
+    }
+
+    pub(super) fn get_block_states(&mut self, hashes: &[Digest]) -> Result<Vec<BlockStatus>> {
+        //todo: optimize this?
+        hashes
+            .iter()
+            .map(|hash| self.get_block_state(hash))
+            .collect::<Result<Vec<_>>>()
+    }
+
+    pub(super) fn get_block(&mut self, hash: &Digest) -> Result<SerialBlock> {
+        let header = self.get_block_header(hash)?;
+        let raw_transactions = self
+            .inner
+            .get(KeyValueColumn::BlockTransactions, &hash[..])?
+            .ok_or_else(|| anyhow!("missing transactions for block"))?;
+        let transactions = SerialBlock::read_transactions(&mut &raw_transactions[..])?;
+        Ok(SerialBlock { header, transactions })
+    }
+
+    pub(super) fn canon_height(&mut self) -> Result<u32> {
+        self.read_meta_u32(KEY_BEST_BLOCK_NUMBER, Some(0))
```

Comment:
is getting the block height from storage fast enough to not care about tracking it in memory?

Reply:
i figured if we ran into a bottleneck, lets do a writethrough cache properly. so far, it seems faster than before though overall. didn't do specific benchmarks on this point -- doesn't seem worth it.